### PR TITLE
Implement library support for service discovery

### PR DESCRIPTION
**Pull request by rynowak**
Implements three flavor of library support for service-discovery:

- GetConnectionString (arbitrary strings) augmenting existing
  functionality already in asp.net core
- GetServiceUri (uris) can be set via connection string or combined
  from a protocol/host/port triple
- GetServiceHost (host/authority) can be set via connection string or
  combined from a host/port pair

This uses the term `host` even though the correct technical term is
`authority`. There are two good reasons for this:

- you should always question authority
- people don't really know that term xD

### CODE REVIEW

**Comment by rynowak** [comment]
Work in progress while I do some:
- [ ] cleanup on samples
- [ ] double-check the tutorial
- [ ] create an issue for breaking changes
- [ ] test that the samples work 👍 
- [ ] write the service discovery docs

File: docs/redis.md [hunk 1]
```diff
@@ -64,7 +64,7 @@ We just showed how `tye` makes it easier to communicate between 2 applications r
 
        services.AddStackExchangeRedisCache(o =>
        {
-            var connectionString = Configuration["connectionstring:redis"];
+            var connectionString = Configuration["connectionstrings:redis"];
```
Comment (davidfowl) [inline docs/redis.md:67]: Use the extension method.

File: src/Microsoft.Tye.Extensions.Configuration/TyeConfigurationExtensions.cs [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+namespace Microsoft.Extensions.Configuration
+{
+    public static class TyeConfigurationExtensions
+    {
+        public static Uri? GetServiceUri(this IConfiguration configuration, string name, string? binding = null)
+        {
+            var key = GetKey(name, binding);
+
+            var connectionString = configuration[$"connectionstrings:{key}"];
+            if (!string.IsNullOrEmpty(connectionString))
```
Comment (davidfowl) [inline src/Microsoft.Tye.Extensions.Configuration/TyeConfigurationExtensions.cs:16]: Is this weird? Feels weird.

File: docs/redis.md (hunk 1 above)
Comment (rynowak) [reply docs/redis.md:67]: image

File: src/Microsoft.Tye.Extensions.Configuration/TyeConfigurationExtensions.cs (hunk 2 above)
Comment (rynowak) [reply src/Microsoft.Tye.Extensions.Configuration/TyeConfigurationExtensions.cs:16]: which part? 

I haven't written the docs yet, but here's what I was going to explain.

> Use the method (`GetConnectionString`, `GetServiceUri`, `GetServiceHost`) **that best describes the kind of config/secret you need to access**.

We need **something** that does this, because in the deployment flow we don't know what code you wrote in your app. Ex: the configuration used to access redis is commonly a hostname/port in development, and commonly a connection string (arbitrary text) in production.

To me it doesn't seem weird once you have the realization that:

- We don't/can't know what code is in your app
- We don't know how to decompose arbitrary text (secrets) into host/port/protocol

---

The potential for mistakes does exist here btw, because imagine that you used `GetServiceHost` for redis. That won't work in production. So we probably need more guidance.

File: docs/redis.md (hunk 1 above)
Comment (davidfowl) [reply docs/redis.md:67]: https://github.com/dotnet/tye/pull/299#issuecomment-609501437

File: src/Microsoft.Tye.Extensions.Configuration/TyeConfigurationExtensions.cs (hunk 2 above)
Comment (rynowak) [reply src/Microsoft.Tye.Extensions.Configuration/TyeConfigurationExtensions.cs:16]: We had a discussion about this more in teams, and ended up with this as the best proposal:

- Don't allow connection strings to override GetServiceUri
- Get rid of GetServiceHost (use GetConnectionString instead for things like redis)
- In dev, we populate CONNECTIONSTRING_REDIS to be the host:port (prepend the protocol if specified)

File: samples/voting/worker/Extensions.cs [hunk 3]
```diff
@@ -10,7 +10,7 @@ public static class Extensions
     {
         public static string GetSqlConnectionString(this IConfiguration configuration)
         {
-            return configuration["connectionstring:postgres"] ?? $"Server={configuration["service:postgres:host"]};Port={configuration["service:postgres:port"]};User Id=postgres;Password=pass@word1;";
+            return configuration["connectionstrings:postgres"] ?? $"Server={configuration["service:postgres:host"]};Port={configuration["service:postgres:port"]};User Id=postgres;Password=pass@word1;";
```
Comment (rynowak) [inline samples/voting/worker/Extensions.cs:13]: @davidfowl - this is something we should decide on...

----

Let me propose an ideal

I think the ideal thing here would be to use templating in the YAML for this.

So you'd write:

```yaml
services:
- name: postgres
  image:  postgres
  env:
    - name: POSTGRES_PASSWORD
      value: "pass@word1"
  bindings:
  - connectionString: "Server=%SERVICE__POSTGRES__HOST%;Port=%SERVICE__POSTGRES__PORT%;User Id=postgres;Password=pass@word1"
```
This doesn't require any tricky code in libraries or your app. It's roughly as simple as what this code is doing, and things like a hardcoded username/password move out of your code. It got **more** maintainable because the hardcoded password isn't used across multiple files.

You could even do better:

```yaml
services:
- name: postgres
  image:  postgres
  env:
    - name: POSTGRES_PASSWORD
      value: "pass@word1"
  bindings:
  - connectionString: "Server=%SERVICE__POSTGRES__HOST%;Port=%SERVICE__POSTGRES__PORT%;User Id=postgres;Password=%POSTGRES_PASSWORD%"
```

We could also enlighten this to use secrets.

---

I think you were onto something with the user having to do some work re: understanding connection strings, but I think it's better if it works in this direction.

File: src/Microsoft.Tye.Hosting/Model/Application.cs [hunk 4]
```diff
@@ -97,10 +97,32 @@ void SetBinding(ServiceDescription targetService, ServiceBinding b)
                     envName = $"{serviceName.ToUpper()}_{b.Name.ToUpper()}";
                 }
 
+                // Use the container name as the host name if there's a single replica (current limitation)
+                var host = b.Host ?? (service.Description.RunInfo is DockerRunInfo ? targetService.Name : defaultHost);
+
+                // Review: should we split this codepath based on the KIND of service that 
+                // provides the binding?
```
Comment (rynowak) [inline src/Microsoft.Tye.Hosting/Model/Application.cs:104]: @davidfowl  Let's discuss this.

I don't really have  strong feeling. What I wonder is if we should **avoid** populating the HOST/PORT env-vars in dev because they won't be there in production. However, the example that I liked about with connection string uses token replacement based on the env-vars, so..... 🤣

File: samples/voting/worker/Extensions.cs (hunk 3 above)
Comment (davidfowl) [reply samples/voting/worker/Extensions.cs:13]: Yea, I'm a fan of this. It might be better if we have a more native syntax for the various pieces:

- Services `${service:name}`
- Env variables `${env:SERVICE__POSTGRES__HOST}`

etc.

File: src/Microsoft.Tye.Hosting/Model/Application.cs (hunk 4 above)
Comment (davidfowl) [reply src/Microsoft.Tye.Hosting/Model/Application.cs:104]: I think it's worth a shot, though if we do container and not container, it'll break in the case where you run with docker.

File: samples/voting/worker/Extensions.cs (hunk 3 above)
Comment (rynowak) [reply samples/voting/worker/Extensions.cs:13]: yeah I agree, that's nice.

Comment (jkotalik) [reply samples/voting/worker/Extensions.cs:13]: Pardon my ignorance, but are we doing anything to consolidate how environment variables value are obtained between windows and linux? Like if I have the string: `Server=%SERVICE__POSTGRES__HOST%`, how would this value be obtained on linux?

**Comment by davidfowl** [comment]
This is looking like a clean coherent story!

**Comment by JamesNK** [comment]
Spicy idea:

Has any thought been given to code generation of known environment variables? Or even just service names. At design/build time we have the yaml file with that info. Some codegen would eliminate magic strings:

```cs
var uri = Configuration.GetServiceUri("backend");

// vs

var uri = Configuration.GetServiceUri(ServiceEnvironments.Backend);
```

**Comment by davidfowl** [comment]
That seems like a fine value add once we have source generators

File: samples/multi-project/tye.yaml [hunk 5]
```diff
@@ -16,4 +16,5 @@ services:
 - name: rabbit
   image: rabbitmq:3-management
   bindings:
-    - port: 5672
+    - protocol: amqp
```
Comment (davidfowl) [inline samples/multi-project/tye.yaml:19]: This isn’t required right?

File: test/Microsoft.Tye.Extensions.Configuration.Tests/TyeConfigurationExtensionsTest.cs [hunk 6]
```diff
@@ -104,21 +104,21 @@ public void GetConnectionString_WithBinding_DoesNotCombineHostAndPort()
...
+                { "service:myservice:protocol", "http" },
+                { "service:myservice:host", "expected.example.com" },
+                { "service:myservice:port", "80" },
             });
 
             var configuration = builder.Build();
 
             var result = configuration.GetServiceUri("myservice");
-            Assert.Equal(new Uri("http://expected.example.com"), result);
+            Assert.Equal(new Uri("http://expected.example.com:80"), result);
```
Comment (davidfowl) [inline test/Microsoft.Tye.Extensions.Configuration.Tests/TyeConfigurationExtensionsTest.cs:121]: Why append 80? Should we have smarts around 443 and 80?

File: samples/mongo-sample/tye.yaml [hunk 7]
```diff
@@ -1,3 +1,4 @@
+name: mongo-sample
```
Comment (davidfowl) [inline samples/mongo-sample/tye.yaml:1]: @jkotalik should this be required?

File: src/Microsoft.Tye.Hosting/PortAssigner.cs [hunk 8]
```diff
@@ -44,12 +44,6 @@ static int GetNextPort()
 
                 foreach (var binding in service.Description.Bindings)
                 {
-                    if (binding.ConnectionString != null)
```
Comment (davidfowl) [inline src/Microsoft.Tye.Hosting/PortAssigner.cs:47]: Why this change?

File: src/Microsoft.Tye.Hosting/Model/Application.cs [hunk 9]
```diff
@@ -80,78 +83,80 @@ string GetValueFromBinding(EnvironmentVariableSource source)
...
-                set($"SERVICE__{configName}__HOST", host);
-                set($"{envName}_SERVICE_HOST", host);
+                set($"SERVICE__{configName}__HOST", binding.Host);
+                set($"{envName}_SERVICE_HOST", binding.Host);
             }
+        }
+
+        // Compute the list of bindings visible to `service`. This is contextual because details like
+        // whether `service` is a container or not will change the result.
+        private List<EffectiveBinding> ComputeBindings(Service service, string defaultHost)
```
Comment (davidfowl) [inline src/Microsoft.Tye.Hosting/Model/Application.cs:131]: not a fan of local functions yet?

**Comment by davidfowl** [comment]
@rynowak are you planning to update the dapr replacement support that was added with the new syntax?

File: docs/frontend_backend_run.md [hunk 10]
```diff
@@ -134,7 +123,13 @@ Now that we have two applications running, let's make them communicate. By defau
     }
    ```
 
-4. Now register this client in `Startup.cs` class in `ConfigureServices` of the `frontend` project:
+4. Add a reference to the `Microsoft.Tye.Extensions.Configuration` package.
+
+    ```txt
+    dotnet add package Microsoft.Tye.Extensions.Configuration
```
Comment (davidfowl) [inline docs/frontend_backend_run.md:129]: Do we need to add the feed until we ship to nuget?

File: src/Microsoft.Tye.Hosting/PortAssigner.cs (hunk 8 above)
Comment (davidfowl) [reply src/Microsoft.Tye.Hosting/PortAssigner.cs:47]: Ah, found it.

File: src/Microsoft.Tye.Hosting/TokenReplacement.cs [hunk 11]
```diff
@@ -0,0 +1,111 @@
...
+                binding = bindings.FirstOrDefault(b => b.Service == keys[1] && b.Name == null);
+                return GetValueFromBinding(binding, keys[2]);
+            }
+            else if (keys.Length == 4 && keys[0] == "service")
+            {
+                binding = bindings.FirstOrDefault(b => b.Service == keys[1] && b.Name == keys[2]);
+                return GetValueFromBinding(binding, keys[3]);
+            }
+
+            return null;
```
Comment (davidfowl) [inline src/Microsoft.Tye.Hosting/TokenReplacement.cs:91]: Throw? Unsupported? We may actually want to validate this syntax during parsing of config so we can give nice line errors (not required for this PR).

**Comment by rynowak** [comment]
> That seems like a fine value add once we have source generators

Yes, this is something we've talked about 👍

File: samples/mongo-sample/tye.yaml (hunk 7 above)
Comment (rynowak) [reply samples/mongo-sample/tye.yaml:1]: Right now the schema says its required, and deployment relies on it. If we made it optional, what would the defualt be? the directory name converted to lower?

File: samples/multi-project/tye.yaml (hunk 5 above)
Comment (rynowak) [reply samples/multi-project/tye.yaml:19]: It is if you want to use `GetServiceUri` for rabbit. Per our discussion:

- We treat everything as URI-like by default
- We use connection strings only when you provided one

The URL that would be returned by `GetServiceUri` would then be `http://localhost:5672` which is not what they accept.


So we could either:

- Show this with a connection string in our examples
- Specify the protocol
- Introduce other methods `GetServiceHost` and `GetServicePort`. 

Obviously I prefer 2, or I would have done one of those other things. Your thoughts?

File: test/Microsoft.Tye.Extensions.Configuration.Tests/TyeConfigurationExtensionsTest.cs (hunk 6 above)
Comment (rynowak) [reply test/Microsoft.Tye.Extensions.Configuration.Tests/TyeConfigurationExtensionsTest.cs:121]: we could, none of the code we have today does this.

File: src/Microsoft.Tye.Hosting/PortAssigner.cs (hunk 8 above)
Comment (rynowak) [reply src/Microsoft.Tye.Hosting/PortAssigner.cs:47]: yeah, we need to assign ports for bindings even when they have a connection string.

File: src/Microsoft.Tye.Hosting/Model/Application.cs (hunk 9 above)
Comment (rynowak) [reply src/Microsoft.Tye.Hosting/Model/Application.cs:131]: This could be a local function but it would close over `this` and so couldn't be static. I wasn't sure how the compiler represents that I just used private.

**Comment by rynowak** [comment]
> are you planning to update the dapr replacement support that was added with the new syntax?

I wasn't planning to make it part of this change, because it's not user-facing, and because I'm running out of time.

File: src/Microsoft.Tye.Hosting/TokenReplacement.cs (hunk 11 above)
Comment (rynowak) [reply src/Microsoft.Tye.Hosting/TokenReplacement.cs:91]: This throws in the caller, which has more context.

File: docs/frontend_backend_run.md (hunk 10 above)
Comment (rynowak) [reply docs/frontend_backend_run.md:129]: Yeah I need to do that in general. Now that we've got a library we need to include that in the instructions everywhere. Good catch.

**Comment by davidfowl** [comment]
Np, make sure you file an issue to follow up.

File: test/Microsoft.Tye.Extensions.Configuration.Tests/TyeConfigurationExtensionsTest.cs (hunk 6 above)
Comment (rynowak) [reply test/Microsoft.Tye.Extensions.Configuration.Tests/TyeConfigurationExtensionsTest.cs:121]: This doesn't actually matter. URI will deal with this for us. If you look at teh codez, you can see that it's wrapped in `new Uri`. I'll change the test.

File: src/Microsoft.Tye.Extensions.Configuration/TyeSecretsConfigurationBuilderExtensions.cs [hunk 12]
```diff
@@ -17,34 +17,26 @@ public static class TyeSecretsConfigurationBuilderExtensions
...
         /// When the environment variable is specified, and the specified directory exists, then the value of
         /// <see cref="TyeSecretsConfigurationSource.FileProvider" /> will be non-null.
         /// </remarks>
-        public static IConfigurationBuilder AddTyeSecrets(this IConfigurationBuilder builder, Action<TyeSecretsConfigurationSource>? configure = null)
+        public static IConfigurationBuilder AddTyeSecrets(this IConfigurationBuilder builder)
         {
-            TyeSecretsConfigurationSource source;
             var secretsDirectory = Environment.GetEnvironmentVariable(TyeSecretsConfigurationSource.TyeSecretsPathEnvironmentVariable);
-            if (secretsDirectory == null || !Directory.Exists(secretsDirectory))
+            if (Directory.Exists(secretsDirectory))
```
Comment (davidfowl) [inline src/Microsoft.Tye.Extensions.Configuration/TyeSecretsConfigurationBuilderExtensions.cs:29]: ?

Comment (rynowak) [reply src/Microsoft.Tye.Extensions.Configuration/TyeSecretsConfigurationBuilderExtensions.cs:29]: This was straight up broken. The config provider was totally busto, because the KPF-file provider doesn't search recursively.

What's here now matches the copy-paste code that we deleted a week agao.

**Comment by rynowak** [comment]
@davidfowl @jkotalik - any more feedback on this one (see the new doc?). I'm trying to wrap up the last few things...

File: samples/mongo-sample/tye.yaml (hunk 7 above)
Comment (davidfowl) [reply samples/mongo-sample/tye.yaml:1]: I just want to make sure it fails if it's not there.

File: samples/multi-project/tye.yaml (hunk 5 above)
Comment (davidfowl) [reply samples/multi-project/tye.yaml:19]: Option 4, default to TCP and the caller would have to know they have to change it to amqp. Either way it seems like its not required but nice to do it here so that the APIs work.

File: samples/mongo-sample/tye.yaml (hunk 7 above)
Comment (rynowak) [reply samples/mongo-sample/tye.yaml:1]: Ok so the answer here is that `ApplicationFactory` will default to the directory name. So it's not required. I need to update the docs and schema.

Do you see a drawback to it being optional? I'm cool with it.

File: samples/multi-project/tye.yaml (hunk 5 above)
Comment (rynowak) [reply samples/multi-project/tye.yaml:19]: It's still important to try and agree on what we're showing. If we do what you're suggesting then when we prompt users at the commandline to get a URI during deployment, they will put in *something* 🤷‍♂. I think it's better in that regard to show that you have control over the URI - which is why I ended up here.

File: samples/mongo-sample/tye.yaml (hunk 7 above)
Comment (jkotalik) [reply samples/mongo-sample/tye.yaml:1]: I'm fine with it being optional and using the directory name as a default.
